feat: add `RasterValue` enum and `get_value_at_pixel` method

vini-fda opened the pull request:
- [x] I agree to follow the project's [code of conduct](https://github.com/georust/.github/blob/main/CODE_OF_CONDUCT.md).
---

This PR proposes to replace the `unwrap_primitive_type` macro with a `RasterValue` enum, which makes it easier to add more types later. It also adds the convenience methods `sample_type`, `get_value_at` and `get_value_at_pixel` to the `GeoTIFF` struct.

The reasoning behind this change is to make a more ergonomic API to add the `CInt32` and `CInt32` data types in another PR (see #24).

vini-fda committed:
feat: added support for Complex Integer types CInt16 & CInt32

vini-fda committed:
fix: added feature flag 'tie-points' to gate the test

vini-fda committed:
chore: minor renaming of types

vini-fda committed:
chore: removed complex int types

vini-fda committed:
chore: roll back changes

vini-fda committed:
feat: re-introduced the get_value_at_pixel method to the GeoTIFF struct

weiji14 commented on Cargo.toml:
```diff
@@ -15,7 +15,7 @@ geo-index = { version = "0.1", optional = true }
 geo-types = { version = "0.7" }
 num_enum = "0.7"
 num-traits = "0.2"
```
Since we're not using `num_traits::FromPrimitive` anymore, I guess this dependency can be removed?
```suggestion
```

weiji14 reviewed (approved):
Sorry for the late review @vini-fda. This PR looks ok to me, a little unfortunate that we can't use the primitive converters from `num-traits` to support complex numbers, but I guess we'll have to settle with more verbose code.

If you remove the `num-traits` dependency and re-lock the Cargo.lock file, I can merge this afterwards. Thanks again!

jurisk commented:
I appreciate you adding `get_value_at_pixel ` for an easier migration path from the `0.0` version. Thank you!

weiji14 commented:
@vini-fda, ok if I push changes directly here to resolve the merge conflict, and then merge this PR in?

Maia20Taurus commented:
This could be really useful, are there any updates on this?

weiji14 committed:
Merge branch 'main' into raster_value

weiji14 committed:
Remove num-traits

weiji14 committed:
fix: import RasterValue in src/lib.rs

weiji14 commented:
Yeah, let me go ahead and handle the merge conflict and fix the tests.

weiji14 committed:
fix: downgrade to Cargo.lock version=3

Fix `lock file version `4` was found, but this version of Cargo does not understand this lock file, perhaps Cargo needs to be updated?`

weiji14 commented:
@Maia20Taurus, do you want to give this a quick review to see if it looks ok to you? Otherwise I'll merge in ~24 hours if there are no objections (or sooner if there's another approval).

Maia20Taurus commented:
No objections, it works for me :)
